test/unittest/rtps/discovery/EdpTests.cpp

```diff
@@ -0,0 +1,476 @@
+    }
+
+    fastdds::dds::PolicyMask writer_incompatible_qos_policies(GUID_t wguid)
+    {
+        return writer_status_[wguid].offered_incompatible_qos;
+    }
+
+};
+
+class EdpTests : public ::testing::Test
+{
+protected:
+
+    void SetUp() override
+    {
+        wdata = new ::testing::NiceMock<WriterProxyData>(1,1);
+        rdata = new ::testing::NiceMock<ReaderProxyData>(1,1);
+
+        //Same Topic
+        wdata->topicName("Topic");
+        rdata->topicName("Topic");
+
+        //Same Topic Kind
+        wdata->topicKind(TopicKind_t::NO_KEY);
+        rdata->topicKind(TopicKind_t::NO_KEY);
+
+        //With no type information, only Type name is compared
+        rdata->m_qos.type_consistency.m_force_type_validation = false;
+        wdata->typeName("TypeName");
+        rdata->typeName("TypeName");
+
+        rdata->isAlive(true);
+        wdata->isAlive(true);
+
+        edp = new EDPMock(&pdp_, &participant_);
+    }
+
+    void TearDown() override
+    {
+        delete edp;
+        delete wdata;
+        delete rdata;
+    }
+
+    void set_incompatible_topic()
+    {
+        rdata->topicName("AnotherTopic");
+    }
+
+    void set_incompatible_topic_kind()
+    {
+        rdata->topicKind(TopicKind_t::WITH_KEY);
+    }
+
+    void set_incompatible_type()
+    {
+        rdata->typeName("AnotherTypeName");
+    }
+
+
+    ::testing::NiceMock<PDP> pdp_;
+    ::testing::NiceMock<RTPSParticipantImpl> participant_;
+    ::testing::NiceMock<WriterProxyData>* wdata;
+    ::testing::NiceMock<ReaderProxyData>* rdata;
+    EDPMock* edp;
+};
+
+
+TEST_F(EdpTests, CompleteCompatibility)
+{
+    EXPECT_TRUE(edp->validMatching(wdata, rdata));
+    EXPECT_TRUE(edp->writer_incompatible_qos_policies(wdata->guid()).none());
+
+    EXPECT_TRUE(edp->validMatching(rdata, wdata));
+    EXPECT_TRUE(edp->reader_incompatible_qos_policies(rdata->guid()).none());
+}
+
+TEST_F(EdpTests, IncompatibleTopic)
+{
+    set_incompatible_topic();
+    EXPECT_FALSE(edp->validMatching(wdata, rdata));
+    EXPECT_TRUE(edp->writer_incompatible_qos_policies(wdata->guid()).none());
+
+    EXPECT_FALSE(edp->validMatching(rdata, wdata));
+    EXPECT_TRUE(edp->reader_incompatible_qos_policies(rdata->guid()).none());
+}
+
+TEST_F(EdpTests, IncompatibleTopicKind)
+{
+    set_incompatible_topic_kind();
+    EXPECT_FALSE(edp->validMatching(wdata, rdata));
+    EXPECT_TRUE(edp->writer_incompatible_qos_policies(wdata->guid()).none());
+
+    EXPECT_FALSE(edp->validMatching(rdata, wdata));
+    EXPECT_TRUE(edp->reader_incompatible_qos_policies(rdata->guid()).none());
+}
+
+TEST_F(EdpTests, IncompatibleType)
+{
+    set_incompatible_type();
+    EXPECT_FALSE(edp->validMatching(wdata, rdata));
+    EXPECT_TRUE(edp->writer_incompatible_qos_policies(wdata->guid()).none());
+
+    EXPECT_FALSE(edp->validMatching(rdata, wdata));
+    EXPECT_TRUE(edp->reader_incompatible_qos_policies(rdata->guid()).none());
+}
+
+TEST_F(EdpTests, CheckDurabilityCompatibility)
```

Comment:
On this and all the following tests, I would follow a pattern of:
* Defining a list of tuples with offered value, requested value, expected result. For the policies depending on an enumeration, I would put all possible combinations of the values.
* Traverse the list assigning the values and calling `check_expectations(expected_result, policy_id)`

We can make an exception on the partitions one, which I would put before this test.